common/threadloop.hpp

```diff
@@ -81,6 +81,7 @@ class threadloop : public plugin {
 		record_coalescer it_log {record_logger_};
 		std::cout << "thread," << std::this_thread::get_id() << ",threadloop," << name << std::endl;
 
+		_m_stoplight->wait_for_ready();
```

Comment:
My mistake. This is removed.

Reply:
I believe we opted to wait before `_p_thread_setup` instead of after in order to fix a crash. @qinjunj Can you confirm this?

Reply:
I just tried and both work fine.

Reply:
We still need `wait_for_ready` before the thread setup because some components will access RelativeClock->`now()` [ex.gldemo](https://github.com/ILLIXR/ILLIXR/blob/issue-58-standardize-clock/gldemo/plugin.cpp#L113) in `p_thread_setup`, which has to be after [RelativeClock->`start()`](https://github.com/ILLIXR/ILLIXR/blob/issue-58-standardize-clock/runtime/runtime_impl.hpp#L65).  `wait_for_ready` guarantees that the RelativeClock is started.

Reply:
That would probably work, but make ILLIXR less deterministic in the `offline_imu_cam` case. The `load_so` time is non-deterministic, which would make a non-deterministic span of time elapse before  `offline_imu_cam`, which tells it which data to submit.


I suggest either:
-  having two `Event`s in `Stoplight` (one for synchronizing `_p_thread_setup()` after `RelativeClock->start()` and one for synchronizing the main loop of each thread.
- Making `RelativeClock->now()` wait (using an `Event`) if it hasn't been started yet. This is more magical, but probably does the right thing most of the time.

Reply:
Also note that `runtime_impl::load_so(const std::string_view)` and `runtime_impl::load_plugin_factory(runtime_factory)` would unintentionally bypass your proposed solution, Huzaifa.

Reply:
What's our decision here? For the two Sam suggested, I think the first one will involve fewer places to make changes, but we don't actually need to sync `_p_thread_setup()`, right? For the second, do we add wait for all `RelativeClock->now()` or just the ones in thread setup? @charmoniumQ Can you explain more why `runtime_impl::load_so(const std::string_view)` and `runtime_impl::load_plugin_factory(runtime_factory)` would unintentionally bypass the start() if we move it to an earlier point in `load_so()`?

Reply:
Given the time constraint, we should let this be for now. Just add a note in the code saying that in the future we need to synchronize the thread mains instead of the thread setups.

Reply:
> Can you explain more why `runtime_impl::load_so(const std::string_view)` (called "method 1") and `runtime_impl::load_plugin_factory(runtime_factory)` (called "method 2") would unintentionally bypass the `start()`?

Putting synchronization in `runtime_impl::load_so(const std::vector<std::string>& so_paths)` (called "method 3") only covers the cases where plugins are loaded by method 3. Note that in the Monado integration, we load plugins with methods 1 and 2, which is bypasses the synchronization in method 3.